moonraker/components/wled.py

```diff
@@ -130,29 +144,52 @@ def _update_color_data(self: Strip,
 
     async def _send_wled_command(self: Strip,
                                  state: Dict[str, Any]) -> None:
+        try:
+            if self.use_http:
+                await self._send_wled_command_http(state)
+            else:
+                await self._send_wled_command_serial(state)
+
+            self.error_state = None
+        except Exception as e:
+            msg = f"WLED: Error {e}"
+            self.error_state = msg
+            logging.exception(msg)
+            raise self.server.error(msg)
+
+    async def _send_wled_command_http(self: Strip,
+                                      state: Dict[str, Any]) -> None:
+        async with self.request_mutex:
+            logging.debug(f"WLED: url:{self.url} json:{state}")
+
+            headers = {"Content-Type": "application/json"}
+            request = HTTPRequest(url=self.url,
+                                  method="POST",
+                                  headers=headers,
+                                  body=json.dumps(state),
+                                  connect_timeout=self.timeout,
+                                  request_timeout=self.timeout)
+            response = await self.client.fetch(request)
+
+            logging.debug(
+                f"WLED: url:{self.url} status:{response.code} "
+                f"response:{response.body}")
+
+    async def _send_wled_command_serial(self: Strip,
+                                        state: Dict[str, Any]) -> None:
         async with self.request_mutex:
-            try:
-                logging.debug(f"WLED: url:{self.url} json:{state}")
-
-                headers = {"Content-Type": "application/json"}
-                request = HTTPRequest(url=self.url,
-                                      method="POST",
-                                      headers=headers,
-                                      body=json.dumps(state),
-                                      connect_timeout=self.timeout,
-                                      request_timeout=self.timeout)
-                response = await self.client.fetch(request)
-
-                logging.debug(
-                    f"WLED: url:{self.url} status:{response.code} "
-                    f"response:{response.body}")
-
-                self.error_state = None
-            except Exception as e:
-                msg = f"WLED: Error {e}"
-                self.error_state = msg
-                logging.exception(msg)
-                raise self.server.error(msg)
+            logging.debug(f"WLED: serial:{self.ser.name} json:{state}")
+
+            if not self.ser.is_open:
+                self.ser.open()
+
+            # asyncio support is still experimental in pySerial
+            self.ser.write(json.dumps(state).encode())
+            self.ser.flush()
```

Comment:
We can't add blocking code to Moonraker.   After opening the serial port we must set the fd to a nonblocking state and we can't call flush().  If wled is sensitive to this it is possible to use asyncio to poll the fd and write to it when its ready.

Reply:
Removed the flush() and changed write_timeout to 0 which I believe is non blocking